### ISSUE TITLE
[Bug]: TRACKING ISSUE: CUDA OOM with Logprobs

### ISSUE THREAD

**Issue by robertgshaw2-redhat** [vllm-project/vllm#5907]
### Your current environment

```text
The output of `python collect_env.py`
```

### 🐛 Describe the bug

vLLM has an issue where we can go OOM if too many `logprobs` are requested.

The reason that this happens is that there are three sources of memory usage:
* Model weights
* KV caches
* Activations

When determining the KV cache size, we calculate peak memory running a long prefill * without logprobs *

If a prompt requests many logprobs, however, this is an additional source of memory usage which is not considered during warmup and can cause OOM because we have nothing in scheduler to prevent this

We have received several examples of this:
* https://github.com/vllm-project/vllm/issues/5890
* https://github.com/vllm-project/vllm/issues/5060 << some of the OOM issues in `AsyncEngineDeadError`
* Anyone running MMLU or ARC in lm-eval-harness

Attempt to fix this:
* https://github.com/vllm-project/vllm/pull/5355

I am working on a design to address this issue

**Comment by neubig** [comment]
Hi @robertgshaw2-neuralmagic , I was wondering if there were any ongoing attempts to resolve this issue, or if #5355 seems like an acceptable fix? Having this fixed would be very helpful to us!

**Comment by binxuan** [comment]
In my case with large vocab size and long seq length, the [logits sort](https://github.com/vllm-project/vllm/blob/3f8d42c81fe8d3842a9e05c9f5d98290b7f79736/vllm/model_executor/layers/sampler.py#L271) is the most memory consuming part and can easily trigger OOM on one single GPU. Is it possible to do some kind of sequence parallel to distribute it to all TP workers?

**Comment by cjfcsjt** [comment]
@robertgshaw2-neuralmagic Any update? Thanks.

**Comment by njhill** [comment]
@tjohnson31415 is looking into this.

**Comment by tjohnson31415** [comment]
I measured the peak GPU memory usage during the processing in `Sampler`. I found that the memory usage is balooned to 9x the size of the input `logits` tensor. The increase comes from upscaling logits to float32 (+2x), copying the tensor for `probs` and `logprobs` (+4x), and processing in `_sample` creating another temporary copy (+2x). I'm sure there are ways to reduce this spike in memory, but we'd still need the input tesnor limited.

The main challenge is if any `prompt_logprobs` are requested since the logits tensor will have logits for every token in the prompt. We will hit limits for a model with a large context and vocab size in the size of the `logits` tensor even before processing in `Sampler`. With Llama 3.1 models with a vocab_size of 128256 and a max sequence length of 131072, a single request with prompt_logprobs requested could produce a logits tensor of 128256 * 131072 * 2B =~ 31 GiB.

Limiting the number of tokens processed at a time with chunked prefill seems like the right solution to me. However, for cases where chunked prefill is not supported, we may still need "chunked logits processing" to limit the number of tokens in the processing from hidden states -> logprobs output. This may be difficult to implement though.

@robertgshaw2-neuralmagic: What do you think about this approach of "chunked logits processing"?

**Comment by robertgshaw2-neuralmagic** [comment]
Is there any fundamental reason we need to make all the copies? Otherwise, it would make sense to me that chunking could work

**Comment by patrickvonplaten** [comment]
Also running into this issue here with small models (2B) and when returning logprobs

**Comment by fzyzcjy** [comment]
Hi, I am seeing the same problem. Is there any updates? Thanks!

**Comment by dangxingyu** [comment]
Experiencing the same issue with the QwQ-32B model on 8xA100 with `max_num_seq=1 `when returning the `prompt_logprobs`.

**Comment by youkaichao** [comment]
> I measured the peak GPU memory usage during the processing in `Sampler`. I found that the memory usage is balooned to 9x the size of the input `logits` tensor. The increase comes from upscaling logits to float32 (+2x), copying the tensor for `probs` and `logprobs` (+4x), and processing in `_sample` creating another temporary copy (+2x). I'm sure there are ways to reduce this spike in memory, but we'd still need the input tesnor limited.
> 
> The main challenge is if any `prompt_logprobs` are requested since the logits tensor will have logits for every token in the prompt. We will hit limits for a model with a large context and vocab size in the size of the `logits` tensor even before processing in `Sampler`. With Llama 3.1 models with a vocab_size of 128256 and a max sequence length of 131072, a single request with prompt_logprobs requested could produce a logits tensor of 128256 * 131072 * 2B =~ 31 GiB.
> 
> Limiting the number of tokens processed at a time with chunked prefill seems like the right solution to me. However, for cases where chunked prefill is not supported, we may still need "chunked logits processing" to limit the number of tokens in the processing from hidden states -> logprobs output. This may be difficult to implement though.
> 
> @robertgshaw2-neuralmagic: What do you think about this approach of "chunked logits processing"?

"chunked logits processing" looks like the ideal solution @tjohnson31415 

Right now, adding `--enable-chunked-prefill` and set `--max-num-batched-tokens` to be the same as `--max-num-seqs` (256 by default) should solve the problem, although it will make the prefill slower.

**Comment by valiantljk** [comment]
any update on this

**Comment by ice-tong** [comment]
I encountered the same issue when attempting to evaluate using `lm_eval + vllm`.

**Comment by lsjlsj5846** [comment]
Same as @ice-tong , any update on this?

**Comment by ai-nikolai** [comment]
similiar issue for this case:

 @robertgshaw2-redhat @ice-tong (and others) - I am encountering the same issue when loading for example: 
`unsloth/Llama-3.3-70B-Instruct-bnb-4bit` on a 80GB A100. Although the output says:
```
INFO 02-15 11:27:26 model_runner.py:1115] Loading model weights took 36.8356 GB
INFO 02-15 11:27:29 worker.py:267] Memory profiling takes 2.39 seconds
INFO 02-15 11:27:29 worker.py:267] the current vLLM instance can use total_gpu_memory (79.15GiB) x gpu_memory_utilization (0.95) = 75.19GiB
```
(I encounter the same issue with smaller quantized models as well).

I am using:
`Initializing a V0 LLM engine (v0.7.2)`

**Comment by ai-nikolai** [comment]
This seems to have worked:
```python
            max_num_seq = 2
            max_memory = 0.9
            max_model_len=1000 #testing if higher values are possible
            tensor_parallel_size = 1
            self.llm = LLM(
                model=model,
                tensor_parallel_size=tensor_parallel_size,
                gpu_memory_utilization=max_memory,
                max_model_len=max_model_len,
                dtype="auto",
                quantization="bitsandbytes", 
                load_format="bitsandbytes",
                enable_chunked_prefill=True,
                max_num_seqs=max_num_seq,
            )
```

**Comment by hypnopump** [comment]
I can still find this issue

**Comment by ffaisal93** [comment]
Facing the same issue. oom error for log likelihood based task. It takes a long time/get's oom for big mmmlu multilingual datasets (eg. mmmlu_bn_bd) and kind of stalls at "Running loglikelihood requests". Frustrating!

**Comment by ai-nikolai** [comment]
@hypnopump @ffaisal93 - what are your exact settings. (i.e. SamplingParams, ModelParams && Hardware spec, e.g. avail. Mem, CUDA Comp && Version of software etc.).

**Generally:**
There is no automatic OOM prevention in vllm so you have to manually set the right params like max_seq_len, max_gpu_util etc.

**Comment by Violettttee** [comment]
i run into this problem even if i set `prompt_logprobs` to 2

**Comment by Violettttee** [comment]
I find that we can try using vllm_v1, which will compute prompt_logprobs by chunk, and you just need to decrease GPU_MEMORY just a little to save some GPU memory for this, and it works out~
Also, it will be more stable if you decrease MAX_BATCH_TOKEN by enabling prefix chunk.
